File: lager1/stadiljus/submissions/partially_accepted/cheese_n2_vis.cpp
// @EXPECTED_GRADES@ AC AC AC TLE TLE
//
// Cheese: Computes ALL O(n^2) visibility pairs, then extracts the Cartesian
// tree from the visibility data. Strictly O(n^2) regardless of input.

#include <bits/stdc++.h>
using namespace std;

const int INF = 1e9;
const int MAXN = 300001;

int p[MAXN], L[MAXN], R[MAXN], par[MAXN];
int ntl[MAXN], ntr[MAXN]; // nearest taller left/right
int dp[MAXN][2][2][2][2];

void solve(int root, int n) {
    vector<int> order;
    order.reserve(n);
    vector<int> stk = {root};
    while (!stk.empty()) {
        int v = stk.back(); stk.pop_back();
        order.push_back(v);
        if (L[v] != -1) stk.push_back(L[v]);
        if (R[v] != -1) stk.push_back(R[v]);
    }

    for (int idx = (int)order.size() - 1; idx >= 0; idx--) {
        int node = order[idx];
        int ln = L[node], rn = R[node];

        for (int el = 0; el < 2; el++)
            for (int er = 0; er < 2; er++)
                for (int hl = 0; hl < 2; hl++)
                    for (int hr = 0; hr < 2; hr++)
                        dp[node][el][er][hl][hr] = INF;

        for (int el = 0; el < 2; el++) {
            for (int er = 0; er < 2; er++) {
                int cost = 1;
                if (ln != -1) {
                    int best = INF;
                    for (int hl = 0; hl < 2; hl++)
                        for (int hr = 0; hr < 2; hr++)
                            best = min(best, dp[ln][el][1][hl][hr]);
                    cost += best;
                }
                if (rn != -1) {
                    int best = INF;
                    for (int hl = 0; hl < 2; hl++)
                        for (int hr = 0; hr < 2; hr++)
                            best = min(best, dp[rn][1][er][hl][hr]);
                    cost += best;
                }
                dp[node][el][er][1][1] = cost;

                struct State { int cost, hl, hr; };
                State ls[4], rs[4];
                int lc = 0, rc = 0;

                if (ln != -1) {
                    for (int hl = 0; hl < 2; hl++)
                        for (int hr = 0; hr < 2; hr++)
                            ls[lc++] = {dp[ln][el][0][hl][hr], hl, hr};
                } else {
                    ls[lc++] = {0, 0, 0};
                }

                if (rn != -1) {
                    for (int hl = 0; hl < 2; hl++)
                        for (int hr = 0; hr < 2; hr++)
                            rs[rc++] = {dp[rn][0][er][hl][hr], hl, hr};
                } else {
                    rs[rc++] = {0, 0, 0};
                }

                for (int a = 0; a < lc; a++) {
                    for (int b = 0; b < rc; b++) {
                        if (ls[a].cost >= INF || rs[b].cost >= INF) continue;
                        if (!(el || er || ls[a].hr || rs[b].hl)) continue;
                        int ol = (ln != -1) ? ls[a].hl : 0;
                        int or_ = (rn != -1) ? rs[b].hr : 0;
                        dp[node][el][er][ol][or_] = min(
                            dp[node][el][er][ol][or_],
                            ls[a].cost + rs[b].cost);
                    }
                }
            }
        }
    }

    int ans = INF;
    for (int hl = 0; hl < 2; hl++)
        for (int hr = 0; hr < 2; hr++)
            ans = min(ans, dp[root][0][0][hl][hr]);
    printf("%d\n", ans);
}

int main() {
    int n;
    scanf("%d", &n);
    for (int i = 0; i < n; i++) scanf("%d", &p[i]);

    memset(L, -1, sizeof(int) * n);
    memset(R, -1, sizeof(int) * n);
    memset(par, -1, sizeof(int) * n);
    memset(ntl, -1, sizeof(int) * n);
    memset(ntr, -1, sizeof(int) * n);

    // O(n^2) visibility scan - no early termination
    for (int i = 0; i < n; i++) {
        int mx = 0;
        for (int j = i + 1; j < n; j++) {
            int mn = min(p[i], p[j]);
            if (mx < mn) {
                if (p[j] > p[i] && ntr[i] == -1)
                    ntr[i] = j;
                if (p[i] > p[j])
                    ntl[j] = i;
            }
            mx = max(mx, p[j]);
        }
    }

    // Build tree from nearest taller
    for (int i = 0; i < n; i++) {
        int lg = ntl[i], rg = ntr[i];
        if (lg == -1) par[i] = rg;
        else if (rg == -1) par[i] = lg;
        else if (p[lg] < p[rg]) par[i] = lg;
        else par[i] = rg;
    }

    int root = -1;
    for (int i = 0; i < n; i++) {
        if (par[i] == -1) root = i;
        else if (i < par[i]) L[par[i]] = i;
        else R[par[i]] = i;
    }

    solve(root, n);
    return 0;
}

File: lager1/stadiljus/submissions/partially_accepted/cheese_n2_vis.py
#!/usr/bin/env python3
# @EXPECTED_GRADES@ AC AC AC TLE TLE
#
# Cheese: Computes ALL O(n^2) visibility pairs, then extracts the Cartesian
# tree from the visibility data. Strictly O(n^2) regardless of input.

import sys

INF = 10**9


def build_cartesian_tree(p):
    n = len(p)
    left = [-1] * n
    right = [-1] * n
    parent = [-1] * n

    nearest_taller_left = [-1] * n
    nearest_taller_right = [-1] * n

    # O(n^2) scan of all pairs - no early termination
    for i in range(n):
        mx = 0
        for j in range(i + 1, n):
            if mx < min(p[i], p[j]):
                # i and j can see each other
                if p[j] > p[i] and nearest_taller_right[i] == -1:
                    nearest_taller_right[i] = j
                if p[i] > p[j]:
                    nearest_taller_left[j] = i
            mx = max(mx, p[j])

    for i in range(n):
        lg = nearest_taller_left[i]
        rg = nearest_taller_right[i]
        if lg == -1:
            parent[i] = rg
        elif rg == -1:
            parent[i] = lg
        elif p[lg] < p[rg]:
            parent[i] = lg
        else:
            parent[i] = rg

    root = -1
    for node, par in enumerate(parent):
        if par == -1:
            root = node
        elif node < par:
            left[par] = node
        else:
            right[par] = node

    return root, left, right


def solve(left, right, root):
    order = []
    stack = [root]
    while stack:
        node = stack.pop()
        order.append(node)
        if left[node] != -1:
            stack.append(left[node])
        if right[node] != -1:
            stack.append(right[node])

    dp = [None] * len(left)
    for node in reversed(order):
        lnode = left[node]
        rnode = right[node]
        cur = [[[[INF] * 2 for _ in range(2)] for _ in range(2)] for _ in range(2)]

        for ext_left in range(2):
            for ext_right in range(2):
                cost = 1
                if lnode != -1:
                    best = INF
                    child = dp[lnode]
                    for has_left in range(2):
                        for has_right in range(2):
                            best = min(best, child[ext_left][1][has_left][has_right])
                    cost += best
                if rnode != -1:
                    best = INF
                    child = dp[rnode]
                    for has_left in range(2):
                        for has_right in range(2):
                            best = min(best, child[1][ext_right][has_left][has_right])
                    cost += best
                cur[ext_left][ext_right][1][1] = cost

                left_states = [(0, 0, 0)]
                if lnode != -1:
                    child = dp[lnode]
                    left_states = []
                    for has_left in range(2):
                        for has_right in range(2):
                            left_states.append(
                                (child[ext_left][0][has_left][has_right], has_left, has_right)
                            )

                right_states = [(0, 0, 0)]
                if rnode != -1:
                    child = dp[rnode]
                    right_states = []
                    for has_left in range(2):
                        for has_right in range(2):
                            right_states.append(
                                (child[0][ext_right][has_left][has_right], has_left, has_right)
                            )

                for left_cost, left_has_left, left_has_right in left_states:
                    for right_cost, right_has_left, right_has_right in right_states:
                        if left_cost >= INF or right_cost >= INF:
                            continue
                        if not (
                            ext_left or ext_right or left_has_right or right_has_left
                        ):
                            continue
                        out_left = left_has_left if lnode != -1 else 0
                        out_right = right_has_right if rnode != -1 else 0
                        cur[ext_left][ext_right][out_left][out_right] = min(
                            cur[ext_left][ext_right][out_left][out_right],
                            left_cost + right_cost,
                        )

        dp[node] = cur

    return min(
        dp[root][0][0][has_left][has_right]
        for has_left in range(2)
        for has_right in range(2)
    )


def main():
    input = sys.stdin.readline
    n = int(input())
    p = list(map(int, input().split()))
    root, left, right = build_cartesian_tree(p)
    print(solve(left, right, root))


if __name__ == "__main__":
    main()
